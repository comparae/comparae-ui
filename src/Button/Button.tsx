import React from 'react';
import styled from 'styled-components';
import { darken } from 'polished';

type Color = 'primary' | 'secondary' | 'error';
type Size = 'small' | 'medium' | 'large';

interface Props {
  /**
   * Set this to change the color
   */
  color: Color;

  /**
   * Disables the button, changing the visual style and make it unable to be pressed
   * @default false
   */
  disabled?: boolean;

  /**
   * The size of the button
   * @default medium
   */
  size?: Size;

  /**
   * The type of the button
   * @default button
   */
  type?: 'submit' | 'button';

  /**
   * Loading state when waiting for an action to complete
   * @default false
   */
  loading?: boolean;

  /** Function to handle button click */
  onClick?(e: React.MouseEvent<HTMLElement>): void;
}

interface StyledProps {
  color: Color;
  size: Size;
  loading?: boolean;
  disabled: boolean;
}

export const Button: React.SFC<Props> = ({
  children,
  disabled = false,
  type = 'button',
  loading = undefined,
  onClick,
  color,
  size = 'medium',
}) => (
  <ButtonStyled
    disabled={disabled}
    type={type}
    onClick={onClick}
    color={color}
    size={size}
    loading={loading}
    data-testid="button"
  >
    {children}
  </ButtonStyled>
);

const ButtonStyled = styled.button<StyledProps>`
  background-color: ${(props) => props.theme.primary};
  border-radius: 4px;
  border: 1px solid ${(props) => props.theme.primary};
  color: #fff;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 16px;
  font-weight: bold;
  height: 40px;
  margin: 0;
  min-width: 140px;
  padding: 0 16px;
  position: relative;
  opacity: ${(props) => (props.disabled ? 0.6 : 1)};
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  transition: border-color 300ms ease, background-color 300ms ease,
    color 300ms ease;
  overflow: hidden;

  &:hover {
    background-color: ${(props) => darken(0.05, props.theme.primary)};
    border-color: ${(props) => darken(0.05, props.theme.primary)};
  }

  &:focus {
    background-color: ${(props) => props.theme.primary};
    border-color: ${(props) => darken(0.1, props.theme.primary)};
  }

  &:active {
    background-color: ${(props) => darken(0.1, props.theme.primary)};
    border-color: ${(props) => darken(0.1, props.theme.primary)};
  }

  &:before {
    content: '';
    position: absolute;
    top: 50%;
    left: 50%;
    display: block;
    width: 0;
    padding-top: 0;
    border-radius: 100%;
    opacity: 0.2;
    background-color: ${(props) => darken(0.1, props.theme.primary)};
    transform: translate(-50%, -50%);
  }

  &:focus:before,
  &:active:before {
    width: 120%;
    padding-top: 120%;
    transition: width 0.4s ease-out, padding-top 0.4s ease-out;
  }
`;
