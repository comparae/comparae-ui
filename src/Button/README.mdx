---
name: Button
route: /button
menu: Components
---

import { Playground, Props } from 'docz';
import { Button } from './Button.tsx';

# Button

Basic Button component

## Usage

<Playground>
  <Button color="primary">Test</Button>
</Playground>

## Props

<Props of={Button} />
