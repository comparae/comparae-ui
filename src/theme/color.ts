export const color = {
  primary: '#81c784',
  secondary: '#6A1B9A',
  grey: '#EEEEEE',
  greyDark: '#52565F',
  red: '#f44336',
};
