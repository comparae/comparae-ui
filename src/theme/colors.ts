export const colors = {
  primary: '#81C784',
  secondary: '#6A1B9A',
  grey: '#EEEEEE',
  greyDark: '#52565F',
  red: '#f44336',
};
