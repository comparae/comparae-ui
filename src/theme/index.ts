export { Dark } from './Dark';
export { Light } from './Light';
export { color } from './color';
