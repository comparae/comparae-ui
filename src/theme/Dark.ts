import { color } from './color';

export const Dark = {
  primary: color.primary,
  background: color.greyDark,
  red: color.red,
};
