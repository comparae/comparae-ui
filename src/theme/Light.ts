import { color } from './color';

export const Light = {
  primary: color.primary,
  background: color.grey,
  red: color.red,
};
