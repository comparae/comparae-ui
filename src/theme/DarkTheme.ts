import { DefaultTheme } from 'styled-components';

import { colors } from './colors';
import { typography } from './typography';

export const DarkTheme: DefaultTheme = {
  borderRadius: '4px',
  colors: {
    primary: colors.primary,
    background: colors.greyDark,
    red: colors.red,
  },
  ...typography,
};
