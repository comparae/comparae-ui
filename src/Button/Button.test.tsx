import * as React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Button } from './Button';
import { Dark, Light } from '../theme';

describe('<Button />', () => {
  it('renders with Light theme', () => {
    const { container } = render(
      <ThemeProvider theme={Dark}>
        <Button color="primary">Test</Button>
      </ThemeProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveStyleRule('color', '#000');
  });

  it('renders with Light theme', () => {
    const { container } = render(
      <ThemeProvider theme={Light}>
        <Button color="primary">Test</Button>
      </ThemeProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveStyleRule('color', '#fff');
  });
});
