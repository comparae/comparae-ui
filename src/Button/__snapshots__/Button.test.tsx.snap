// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders with Light theme 1`] = `
.c0 {
  background-color: #81C784;
  border-radius: 4px;
  border: 1px solid #81C784;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 16px;
  font-weight: bold;
  height: 40px;
  margin: 0;
  min-width: 140px;
  padding: 0 16px;
  position: relative;
  opacity: 1;
  cursor: pointer;
  -webkit-transition: border-color 300ms ease,background-color 300ms ease, color 300ms ease;
  transition: border-color 300ms ease,background-color 300ms ease, color 300ms ease;
  overflow: hidden;
}

.c0:hover {
  background-color: #6fbf73;
  border-color: #6fbf73;
}

.c0:focus {
  background-color: #81C784;
  border-color: #5eb762;
}

.c0:active {
  background-color: #5eb762;
  border-color: #5eb762;
}

.c0:before {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  display: block;
  width: 0;
  padding-top: 0;
  border-radius: 100%;
  opacity: 0.2;
  background-color: #5eb762;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c0:focus:before,
.c0:active:before {
  width: 120%;
  padding-top: 120%;
  -webkit-transition: width 0.4s ease-out,padding-top 0.4s ease-out;
  transition: width 0.4s ease-out,padding-top 0.4s ease-out;
}

<button
  class="c0"
  color="primary"
  data-testid="button"
  type="button"
>
  Test
</button>
`;

exports[`<Button /> renders with Light theme 2`] = `
.c0 {
  background-color: #81C784;
  border-radius: 4px;
  border: 1px solid #81C784;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 16px;
  font-weight: bold;
  height: 40px;
  margin: 0;
  min-width: 140px;
  padding: 0 16px;
  position: relative;
  opacity: 1;
  cursor: pointer;
  -webkit-transition: border-color 300ms ease,background-color 300ms ease, color 300ms ease;
  transition: border-color 300ms ease,background-color 300ms ease, color 300ms ease;
  overflow: hidden;
}

.c0:hover {
  background-color: #6fbf73;
  border-color: #6fbf73;
}

.c0:focus {
  background-color: #81C784;
  border-color: #5eb762;
}

.c0:active {
  background-color: #5eb762;
  border-color: #5eb762;
}

.c0:before {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  display: block;
  width: 0;
  padding-top: 0;
  border-radius: 100%;
  opacity: 0.2;
  background-color: #5eb762;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c0:focus:before,
.c0:active:before {
  width: 120%;
  padding-top: 120%;
  -webkit-transition: width 0.4s ease-out,padding-top 0.4s ease-out;
  transition: width 0.4s ease-out,padding-top 0.4s ease-out;
}

<button
  class="c0"
  color="primary"
  data-testid="button"
  type="button"
>
  Test
</button>
`;
